protocol/morse/sanctioned_endpoints_store.go

```diff
@@ -0,0 +1,190 @@
+	// These sanctions automatically expire after defaultSessionSanctionExpiration
+	sessionSanctions *cache.Cache
+}
+
+// newSanctionedEndpointsStore creates a new sanctionedEndpointsStore
+func newSanctionedEndpointsStore(logger polylog.Logger) *sanctionedEndpointsStore {
+	return &sanctionedEndpointsStore{
+		logger:             logger,
+		permanentSanctions: make(map[protocol.EndpointAddr]sanction),
+		sessionSanctions:   cache.New(defaultSessionSanctionExpiration, defaultCacheCleanupInterval),
+	}
+}
+
+// ApplyObservations processes all observations in the list and applies appropriate sanctions.
+// This is the main public method of the sanctioned endpoints store that should be called by external components
+// to handle observations and apply sanctions as needed.
+func (ses *sanctionedEndpointsStore) ApplyObservations(morseObservations []*protocolobservations.MorseRequestObservations) {
+	if len(morseObservations) == 0 {
+		ses.logger.With("method", "ApplyObservations").Warn().Msg("Skipping processing: received empty observation list")
+		return
+	}
+
+	// Process each observation set
+	for _, observationSet := range morseObservations {
+		// Process each endpoint observation in the set
+		for _, endpointObservation := range observationSet.GetEndpointObservations() {
+			logger := loggerWithEndpointObservation(ses.logger, endpointObservation)
+			logger.Debug().Msg("processing endpoint observation.")
+
+			// Skip if no sanction is recommended
+			recommendedSanction := endpointObservation.GetRecommendedSanction()
+			if recommendedSanction == protocolobservations.MorseSanctionType_MORSE_SANCTION_UNSPECIFIED {
+				continue
+			}
+
+			// Create sanction based on the observation
+			sanctionData := buildSanctionFromObservation(endpointObservation)
+
+			// Apply the appropriate type of sanction
+			switch recommendedSanction {
+
+			// Permanent sanction: these persist for the lifetime of the process but are not persisted to disk.
+			// They will be lost on PATH process restart and are not shared across multiple instances.
+			case protocolobservations.MorseSanctionType_MORSE_SANCTION_PERMANENT:
+				logger.Info().Msg("Adding permanent sanction for endpoint")
+
+				ses.addPermanentSanction(
+					protocol.EndpointAddr(endpointObservation.GetEndpointAddr()),
+					sanctionData,
+				)
+
+			// Session-based sanctions: These automatically expire after a set duration.
+			case protocolobservations.MorseSanctionType_MORSE_SANCTION_SESSION:
+				logger.Info().Msg("Adding session sanction for endpoint")
+
+				ses.addSessionSanction(
+					protocol.EndpointAddr(endpointObservation.GetEndpointAddr()),
+					endpointObservation.GetAppAddress(),
+					endpointObservation.GetSessionKey(),
+					sanctionData,
+				)
+			default:
+				logger.Warn().Msg("sanction type not supported by the store. skipping.")
+			}
+		}
+	}
+}
+
+// FilterSanctionedEndpoints removes sanctioned endpoints from the provided list.
```

Comment:
Confirming this needs to be publically exposed?

Reply:
Yes, not strictly required on a package-private struct, but helps increase readability by indicating what method(s) are to be called by other structs.